<div class="alert alert-block alert-success"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

¡Hola Rodrigo!

Mi nombre es Juan Tellez. Un gusto conocerte, seré tu revisor en este proyecto.

Cuando encuentre un error por primera vez, simplemente lo señalaré para que puedas identificarlo y corregirlo por tu cuenta. Además, a lo largo del notebook iré dejando algunas observaciones para mejorar tu código, fortalecer tu análisis y comentar también tus interpretaciones sobre el tema.

Si en esta primera orientación aún no logras resolver algún punto, en la próxima iteración te dejaré una pista más precisa y, cuando sea necesario, algunos ejemplos prácticos. La idea es que este proceso no sea solo una corrección, sino también una oportunidad para que desarrolles criterio como analista de datos o científico de datos.

También estaré abierto a comentarios y discusiones sobre el tema.

Encontrarás mis comentarios a continuación. Por favor, no los muevas, modifiques ni elimines.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como estos:

</div> <div class="alert alert-block alert-success"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Éxito. Todo se ha hecho de forma exitosa.

</div> <div class="alert alert-block alert-warning"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Observación. Algunas recomendaciones o ajustes no bloqueantes.

</div> <div class="alert alert-block alert-danger"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos.

</div> <div class="alert alert-block alert-info"> <b>Respuesta de estudiante.</b> <a class="tocSkip"></a><br><br> </div>

<div class="alert alert-block alert-danger"> <b>Review General. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Rodrigo, quería tomarme este espacio al inicio de tu proyecto para comentarte mis apreciaciones generales de esta iteración de tu entrega.

Siempre me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes. Te deseo lo mejor en este camino y espero que consigas lograr tus objetivos. Personalmente me gusta compartir este consejo: está bien equivocarse. Es normal y, de hecho, es una de las mejores formas de aprender. Los errores te ayudan a descubrir detalles que no siempre se ven a primera vista, y esa experiencia es la que poco a poco te va formando como un mejor analista o científico de datos.

Ahora sí, yendo a esta entrega:

Tu proyecto está muy bien encaminado. Se nota una estructura ordenada, con una narrativa clara desde la carga de datos hasta el dashboard. También se observa un esfuerzo importante por explicar cada etapa del análisis, no solo por ejecutar código. Eso es valioso porque acerca el notebook a una entrega profesional, donde el lector puede entender qué se hizo, por qué se hizo y cómo se conecta con el negocio.

Como fortalezas principales, cargaste correctamente los datasets principales, trabajaste con copias limpias, eliminaste duplicados en órdenes, convertiste fechas, corregiste la capitalización de país y recuperaste valores faltantes de canal usando una relación lógica con id_campaña. Además, el análisis de rentabilidad tiene varios KPIs útiles, el test A/B está muy bien planteado y el dashboard tiene una presentación bastante completa: incluye KPIs, filtros, evolución mensual, acumulado, análisis por categoría/producto y una página de detalle.

Los puntos más importantes a mejorar están en tres frentes. Primero, conviene dejar más evidencia técnica de algunas validaciones de calidad de datos, especialmente nulos, duplicados y valores inválidos en los tres datasets. Segundo, hay que alinear mejor la definición de profit y las unidades comunicadas en la interpretación, porque en algunas partes se mezcla utilidad bruta con utilidad después de marketing. Tercero, la sección de retención por cohortes sí necesita corrección, ya que el denominador de usuarios iniciales y la lógica semanal pueden cambiar bastante la lectura real de retención.

En general, el proyecto muestra muy buen esfuerzo y una intención clara de conectar análisis con negocio. El dashboard está por encima de lo esperado para una entrega inicial y comunica bien los resultados principales. El ajuste más importante para poder cerrar el proyecto con seguridad está en la retención por cohortes.

Éxitos Rodrigo.

Saludos.

</div>

# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

# Paso 1. Preparación y calidad de datos

### Objetivo:
Evaluar la calidad de los datos, detectar inconsistencias, corregir problemas identificados y generar datasets confiables para los análisis posteriores.

# 1.1 Carga y validación inicial de los datasets

In [1]:
# importar librerías
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from statsmodels.stats.proportion import proportions_ztest

In [2]:
# configurar las visualizaciones en jupyter
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

In [3]:
# cargar archivos
orders = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv')
catalog = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv')
marketing = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv')

In [4]:
# creacion de copias para comparaciones posteriores
orders_clean = orders.copy()
catalog_clean = catalog.copy()
marketing_clean = marketing.copy()

## 1.2 Exploración inicial de los datasets

In [5]:
# explorar datasets
print("Orders")
print(orders_clean.shape)
orders_clean.info()
print(" ")
print("Catalog")
print(catalog_clean.shape)
catalog_clean.info()
print(" ")
print("Marketing")
print(marketing_clean.shape)
marketing_clean.info()

Orders
(25100, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB
 
Catalog
(7, 4)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column      

<div class="alert alert-block alert-success"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Buen inicio, Rodrigo. En esta parte cargaste correctamente los tres datasets principales del proyecto (orders, catalog y marketing) y además creaste copias de trabajo (orders_clean, catalog_clean, marketing_clean) antes de modificar los datos.

Ese flujo es una buena práctica profesional porque preserva los datos originales y permite comparar contra la versión limpia si aparece alguna diferencia en los KPIs más adelante. También es positivo que hayas usado .info() para revisar estructura, tipos de datos y conteo de valores no nulos en los tres datasets.

</div>

# Comentarios iniciales

Los tres datasets fueron cargados correctamente desde las fuentes proporcionadas y se crearon copias de trabajo para preservar los datos originales durante el proceso de limpieza y transformación.

## Observaciones preliminares

- El dataset Orders contiene el mayor volumen de registros y variables, por lo que será la principal fuente para el análisis de negocio.
- Durante la exploración inicial se identificó la presencia de valores faltantes en algunas columnas de Orders y Marketing.
- Se observaron posibles inconsistencias en ciertas variables categóricas del catálogo que requerirán validación durante el análisis de calidad.
- Algunas variables numéricas presentan valores negativos y valores extremos que deberán revisarse para determinar si representan errores o comportamientos válidos del negocio.

---

### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.

Se revisan los 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verificar consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 

---

In [6]:
print("Orders original:", orders.shape)

Orders original: (25100, 12)


In [7]:
orders_clean.duplicated().sum()

100

In [8]:
orders_clean = orders_clean.drop_duplicates()

In [9]:
orders_clean.duplicated().sum()

0

In [10]:
orders_clean["fecha_hora_pedido"] = pd.to_datetime(orders_clean['fecha_hora_pedido'])

In [11]:
orders['pais'].value_counts(dropna=False)

Colombia     7520
Mexico       7502
Argentina    7291
mexico        865
colombia      823
argentina     799
NaN           300
Name: pais, dtype: int64

In [12]:
orders_clean['pais'] = orders_clean['pais'].str.title()

In [13]:
orders_clean['pais'].value_counts(dropna=False)

Mexico       8341
Colombia     8304
Argentina    8055
NaN           300
Name: pais, dtype: int64

In [14]:
marketing_clean['fecha'] = pd.to_datetime(marketing_clean['fecha'])

In [15]:
marketing['canal'].isna().sum()

101

In [16]:
orders_clean['fecha_hora_pedido'].dtype
marketing_clean['fecha'].dtype

dtype('<M8[ns]')

In [17]:
diccionario_canal = marketing_clean[marketing_clean['canal'].notna()].set_index('id_campaña')['canal'].to_dict()

In [18]:
marketing_clean['id_campaña'].map(diccionario_canal)

0           organic
1       paid_search
2            social
3           organic
4       paid_search
           ...     
1615    paid_search
1616         social
1617        organic
1618    paid_search
1619         social
Name: id_campaña, Length: 1620, dtype: object

In [19]:
marketing_clean['canal'] = marketing_clean['canal'].fillna(marketing_clean['id_campaña'].map(diccionario_canal))

Se identificó una relación determinística entre id_campaña y canal, lo que permitió recuperar los valores faltantes sin recurrir a imputaciones arbitrarias.

In [20]:
marketing_clean['canal'].isna().sum()

0

In [21]:
orders_clean.shape

(25000, 12)

# Principales acciones realizadas:

- Se eliminaron 100 registros duplicados en Orders.
- Se estandarizó la variable país para corregir inconsistencias de capitalización.
- Se convirtieron las variables de fecha al formato datetime.
- Se validó la consistencia de montos para los registros con información completa.
- Se recuperaron 101 valores faltantes en la variable canal utilizando la relación determinística entre id_campaña y canal.
- No se encontraron inconsistencias relevantes en Catalog.
- Se identificaron 4 registros con cantidades y montos negativos.
- Tras revisar su consistencia matemática y comportamiento dentro del dataset, se concluyó que podrían representar devoluciones o ajustes válidos del negocio, por lo que fueron conservados.

---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [22]:
# exportar datasets
orders_clean.to_csv('orders_clean.csv', index=False)
catalog_clean.to_csv('catalog_clean.csv', index=False)
marketing_clean.to_csv('marketing_clean.csv', index=False)

La revisión de calidad permitió validar la confiabilidad de los tres datasets antes de iniciar el análisis de negocio. Las inconsistencias identificadas fueron corregidas cuando existía evidencia suficiente para hacerlo, mientras que los registros atípicos o con información parcial fueron conservados tras verificar que podrían representar comportamientos legítimos del negocio. Esto permite avanzar a la etapa de análisis con una base de datos consistente, minimizando el riesgo de sesgos derivados de errores de calidad de datos.

<div class="alert alert-block alert-warning"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

La limpieza va bien encaminada: eliminaste 100 duplicados en orders_clean, convertiste fecha_hora_pedido y fecha a formato datetime, estandarizaste pais y recuperaste los valores faltantes de canal usando la relación con id_campaña.

Para fortalecer esta sección, te recomiendo dejar evidencia más completa de tres revisiones antes de exportar los CSV limpios: nulos por dataset, duplicados por dataset y validación de valores inválidos en variables numéricas clave como cantidad, precio_unitario, monto_total, gasto y costo_unitario.

Mencionas en la explicación que encontraste registros negativos y que podrían representar devoluciones o ajustes válidos del negocio. Esa interpretación puede ser razonable, pero conviene mostrar el filtro o tabla donde se observan esos casos. En un contexto profesional, esta trazabilidad es importante porque permite defender por qué se conservaron esos registros y cómo podrían afectar revenue, costos o rentabilidad.

</div>

---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.

Se usan los 3 datasets (`orders`, `catalog`, `marketing_spend`):

**📊 Parte 1: Rentabilidad del negocio**
- ¿Cuál es el ingreso total (revenue)? 
- ¿Cuál es el costo total? 
- ¿Cuánto se ha invertido en marketing? 
- ¿El negocio es rentable? (calcular profit)  

---

**📈 Parte 2: Comportamiento de ventas**
- ¿Cuál es el ticket promedio por orden? 
- ¿Cuál es la cantidad promedio de productos por orden? 
- ¿Cuál es el producto más vendido?
- ¿Cuánto se ha gastado en marketing por canal? 

In [23]:
orders_clean = orders_clean.merge(catalog_clean, on='nombre_producto', how = 'left')

Se realizó la integración entre Orders y Catalog mediante nombre_producto. La validación posterior confirmó que no se perdieron registros ni revenue durante el proceso. Únicamente 30 registros permanecieron sin costo asociado debido a la ausencia de información de producto identificada previamente durante la fase de limpieza.

In [24]:
registros_sin_costo = orders_clean[orders_clean['costo_unitario'].isna()]

registros_sin_costo['monto_total'].sum()

12262.619999999999

Los 30 registros sin información de costo representan únicamente el 0.023% del revenue total. Debido a que no es posible calcular su costo ni su rentabilidad de forma confiable, fueron excluidos del análisis sin afectar significativamente los resultados globales.

In [25]:
orders_profit = orders_clean[orders_clean['costo_unitario'].notna()].copy()

In [26]:
orders_profit['costo_total'] = (orders_profit['cantidad'] * orders_profit['costo_unitario'])

In [27]:
orders_profit['profit_total'] = (orders_profit['monto_total'] - orders_profit['costo_total'])

In [28]:
orders_profit[['monto_total', 'costo_total', 'profit_total']].head()

,monto_total,costo_total,profit_total
0,665.37,378.62,286.75
1,171.86,25.21,146.65
2,195.99,353.28,-157.29
3,242.87,25.21,217.66
4,336.28,176.64,159.64


In [29]:
revenue_total = orders_profit['monto_total'].sum()

In [30]:
costo_total = orders_profit['costo_total'].sum()

In [31]:
profit_total = orders_profit['profit_total'].sum()

In [32]:
margen = (profit_total / revenue_total)*100

In [33]:
inversion_marketing = marketing_clean['gasto'].sum()

In [34]:
print(revenue_total)
print(costo_total)
print(profit_total)
print(margen)
print(inversion_marketing)
print(profit_total - inversion_marketing)

51975199.63999999
43124018.41
8829553.229999999
16.98801215032716
2871843.53
5957709.699999999


El análisis de los indicadores clave muestra que el negocio generó ingresos totales por $5.20 millones, con costos de producto cercanos a $4.31 millones y una inversión en marketing de $2.87 millones. Considerando únicamente los costos de producto, la operación obtuvo una ganancia de aproximadamente $883 mil, equivalente a un margen de rentabilidad del 16.99%. Estos resultados indican que las ventas generan un beneficio positivo sobre el costo de los productos, aunque será necesario profundizar en el análisis para evaluar el impacto de la inversión en marketing sobre la rentabilidad global del negocio.

<div class="alert alert-block alert-warning"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Aquí los cálculos principales están bastante avanzados: calculaste revenue, costos, utilidad antes de marketing, inversión de marketing y también mostraste el resultado de restar marketing a la utilidad.

El punto a ajustar está en la definición y comunicación de profit. En el código, profit_total se calcula como monto_total - costo_total, es decir, utilidad bruta o margen después de costos de producto. Luego imprimes profit_total - inversion_marketing, que se acerca más a la fórmula esperada para profit total según la rúbrica: revenue menos costos menos marketing.

Además, en la interpretación se mencionan ingresos por $5.20 millones y costos por $4.31 millones, pero el output muestra aproximadamente $51.98M de revenue y $43.12M de costos. Esta diferencia de unidades puede confundir mucho en una presentación ejecutiva.

Para fortalecer esta parte, te recomiendo separar claramente las métricas:
<br><br>

utilidad bruta: revenue menos costos de producto;
profit total o utilidad neta del proyecto: revenue menos costos menos marketing;
margen: indicar si corresponde a margen bruto o margen después de marketing.

Este ajuste es importante porque el dashboard también muestra “Utilidad Total” cercana a $8.9M, que parece corresponder a utilidad bruta. Si el negocio espera profit total según la rúbrica, esa tarjeta debería descontar marketing o, al menos, cambiar el nombre para que no haya ambigüedad.

</div>

In [35]:
orders_profit['monto_total'].mean()

2081.505792551061

In [36]:
orders_profit['cantidad'].mean()

7.121869983948636

In [37]:
orders_profit.groupby('nombre_producto')['cantidad'].sum().sort_values(ascending=False)

nombre_producto
Laptop-Gaming-16GB      144198.0
Vacuum-Pro-Black          6284.0
Blender-XL-Red            6279.0
Jacket-Winter-M           6256.0
Sneakers-Urban-42         6172.0
Tablet-Standard-64GB      4153.0
Phone-Pro-128GB           4135.0
Name: cantidad, dtype: float64

In [38]:
orders_profit.groupby('nombre_producto')['monto_total'].sum().sort_values(ascending=False)

nombre_producto
Laptop-Gaming-16GB      43431270.21
Vacuum-Pro-Black         1623994.00
Blender-XL-Red           1613576.87
Jacket-Winter-M          1610784.01
Sneakers-Urban-42        1574713.00
Tablet-Standard-64GB     1063771.00
Phone-Pro-128GB          1057090.55
Name: monto_total, dtype: float64

In [39]:
marketing_clean.groupby('canal')['gasto'].sum().sort_values(ascending=False)

canal
social         976818.37
organic        972650.96
paid_search    922374.20
Name: gasto, dtype: float64

In [40]:
orders_clean['cantidad'].sum()

177522.0

## Hallazgos clave
- El negocio es rentable.
- El margen bruto supera el 16%.
- El producto más vendido por volumen de piezas es Laptop-GAming-16GB.
- El canal con mayor inversión es el social con $976,818.

<div class="alert alert-block alert-success"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Buen trabajo complementando los KPIs generales con análisis agregados. En esta sección identificaste el producto con mayor cantidad vendida, el producto con mayor revenue y el gasto de marketing por canal.

Esto mejora el análisis porque no se queda únicamente en métricas globales. Por ejemplo, detectar que Laptop-Gaming-16GB concentra una gran parte de la cantidad vendida y del revenue ayuda a conectar los resultados con decisiones de inventario, pricing, campañas o priorización comercial. También es útil ver que el canal social concentra la mayor inversión de marketing, porque permite iniciar una conversación sobre eficiencia por canal.

</div>

---

## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

**📊 Parte 1: Construcción del funnel**
- ¿Cuántos usuarios llegan a cada etapa del funnel?  
- Se calcula el número de usuarios únicos por `nombre_evento`  
- Se ordenan los eventos según el flujo del usuario  

---

**📉 Parte 2: Análisis de conversión**
- Se calcula la tasa de conversión entre cada paso del funnel  
- Se identifica en qué etapa se pierde la mayor cantidad de usuarios  
- ¿Cuál es la tasa de conversión final?
---

In [41]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [42]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head()

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica


In [43]:
events['nombre_evento'].value_counts()

first_visit         29957
add_to_cart         24157
select_item         23887
begin_checkout      17971
add_payment_info    12018
purchase            12010
Name: nombre_evento, dtype: int64

El evento más frecuente es first_visit, seguido por acciones relacionadas con la exploración y compra de productos. Sin embargo, para construir correctamente el funnel será necesario analizar usuarios únicos por etapa, ya que un mismo usuario puede generar múltiples eventos durante su recorrido.

In [44]:
# PARTE 1: Totales del funnel
# ======================

query_totals = '''

SELECT
    nombre_evento,
    COUNT(DISTINCT id_usuario) AS usuarios,
CASE
    WHEN nombre_evento = 'first_visit' THEN 1
    WHEN nombre_evento = 'select_item' THEN 2
    WHEN nombre_evento = 'add_to_cart' THEN 3
    WHEN nombre_evento = 'begin_checkout' THEN 4
    WHEN nombre_evento = 'add_payment_info' THEN 5
    WHEN nombre_evento = 'purchase' THEN 6
END AS orden
FROM events
GROUP BY nombre_evento, orden
ORDER BY orden ASC;


'''

totals = pd.read_sql(query_totals, con=engine)
totals



,nombre_evento,usuarios,orden
0,first_visit,7796,1
1,select_item,7582,2
2,add_to_cart,7634,3
3,begin_checkout,7208,4
4,add_payment_info,6250,5
5,purchase,6240,6


- Se construyó el funnel de conversión utilizando usuarios únicos por evento. La etapa con mayor alcance fue ***first_visit***, con 7,796 usuarios, mientras que 6,240 usuarios completaron una compra. En general, se observa una disminución gradual de usuarios a medida que avanzan por las distintas etapas del proceso de compra, lo que indica que existen pérdidas de usuarios a lo largo del recorrido hasta la conversión final.
- Se observa una conversión superior al 100% entre select_item y add_to_cart. Esto sugiere que algunos usuarios llegaron directamente al carrito sin registrar previamente el evento select_item, una situación común cuando existen rutas alternativas dentro de la aplicación o eventos faltantes en la instrumentación.

In [45]:

# PARTE 2: Conversiones
# ======================

query_conversion = '''
WITH cte_first_visit as (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'first_visit'
),
cte_select_item as (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'select_item'
),
cte_add_to_cart as (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'add_to_cart'
),
cte_begin_checkout as (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'begin_checkout'
),
cte_add_payment_info as (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'add_payment_info'
),
cte_purchase as (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'purchase'
)
SELECT
    (SELECT COUNT(*) FROM cte_first_visit) AS first_visit_users,

    (SELECT COUNT(*) FROM cte_select_item) AS select_item_users,
    (SELECT COUNT(*) FROM cte_add_to_cart) AS add_to_cart_users,
    (SELECT COUNT(*) FROM cte_begin_checkout) AS begin_checkout_users,
    (SELECT COUNT(*) FROM cte_add_payment_info) AS add_payment_info_users,
    (SELECT COUNT(*) FROM cte_purchase) AS purchase_users,

    ROUND((SELECT COUNT(*) FROM cte_select_item) * 100.00 / NULLIF((SELECT COUNT(*) FROM cte_first_visit), 0), 2) AS conversion_select_from_first,
    ROUND((SELECT COUNT(*) FROM cte_add_to_cart) * 100.00 / NULLIF((SELECT COUNT(*) FROM cte_select_item), 0), 2) AS conversion_cart_from_select,
    ROUND((SELECT COUNT(*) FROM cte_begin_checkout) * 100.00 / NULLIF((SELECT COUNT(*) FROM cte_add_to_cart), 0), 2) AS conversion_checkout_from_cart,
    ROUND((SELECT COUNT(*) FROM cte_add_payment_info) * 100.00 / NULLIF((SELECT COUNT(*) FROM cte_begin_checkout), 0), 2) AS conversion_payment_from_checkout,
    ROUND((SELECT COUNT(*) FROM cte_purchase) * 100.00 / NULLIF((SELECT COUNT(*) FROM cte_add_payment_info), 0), 2) AS conversion_purchase_from_payment,
    ROUND((SELECT COUNT(*) FROM cte_purchase) * 100.00 / NULLIF((SELECT COUNT(*) FROM cte_first_visit), 0), 2) AS conversion_purchase_from_first;
    
    

'''


conversion = pd.read_sql(query_conversion, con=engine)
conversion

,first_visit_users,select_item_users,add_to_cart_users,begin_checkout_users,add_payment_info_users,purchase_users,conversion_select_from_first,conversion_cart_from_select,conversion_checkout_from_cart,conversion_payment_from_checkout,conversion_purchase_from_payment,conversion_purchase_from_first
0,7796,7582,7634,7208,6250,6240,97.26,100.69,94.42,86.71,99.84,80.04


El análisis del funnel muestra una conversión final de 80.04% desde first_visit hasta purchase. La mayor pérdida de usuarios ocurre entre las etapas begin_checkout y add_payment_info, donde la conversión es de 86.71%, lo que sugiere posibles fricciones durante el proceso de pago. Asimismo, se observa una conversión superior al 100% entre select_item y add_to_cart, lo que podría indicar diferencias en el registro de eventos o recorridos alternativos de los usuarios dentro de la plataforma. Aunque la conversión final es elevada (80.04%), la mayor oportunidad de optimización se encuentra entre las etapas de exploración y selección de productos, donde ocurre la mayor pérdida de usuarios.

<div class="alert alert-block alert-warning"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

La sección del funnel tiene una base sólida: usaste SQL, construiste CTEs y contaste usuarios únicos por etapa, lo cual es mucho mejor que contar eventos totales.

El punto a mejorar está en la lógica de avance entre etapas. Actualmente cada CTE toma usuarios únicos por evento de forma independiente. Por eso aparece una conversión superior al 100% entre select_item y add_to_cart. Esa señal indica que las etapas no están condicionadas de forma secuencial, es decir, no necesariamente estás midiendo usuarios que pasaron por la etapa anterior antes de avanzar a la siguiente.

En un contexto de negocio, esto puede cambiar la interpretación de la caída principal. Si el funnel no controla el avance secuencial, la pérdida entre etapas puede estar mezclando recorridos alternativos, eventos faltantes o usuarios que entran al flujo por caminos distintos.

Para fortalecer esta parte, te recomiendo ajustar la consulta para que cada etapa se calcule sobre usuarios que ya pasaron por la etapa previa, o dejar muy claro que se trata de un funnel por presencia de evento y no de un funnel estrictamente secuencial. Con eso, la identificación de la mayor caída quedará más confiable.

</div>

---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users` 
- `user_activity` 

---
1. Se identifica la cohorte de cada usuario según el **mes de registro**.


2. Se calcula la retención semanal: cuántos usuarios **se mantienen activos** en cada semana desde su registro.
   - `retenido_w1`: usuarios activos en la semana 1  
   - `retenido_w2`: usuarios activos en la semana 2  
   - `retenido_w3`: usuarios activos en la semana 3  


3. Se calcula el porcentaje de retención para cada semana, dividiendo los usuarios retenidos entre los clientes iniciales de la cohorte:  
   - `semana_1`: porcentaje de usuarios retenidos en la semana 1  
   - `semana_2`: porcentaje de usuarios retenidos en la semana 2  
   - `semana_3`: porcentaje de usuarios retenidos en la semana 3  

Se revisa que la columna de fecha esté en formato correcto (`DATE`).  
Se realiza la conversión usando: `CAST(fecha_registro AS DATE)`

In [46]:
# Explorar tabla users
# =========================
query_users = '''
SELECT *
FROM users;
'''
users = pd.read_sql(query_users, con=engine)
users.head(3)

,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free


In [47]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT *
FROM user_activity;
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.head(3)

,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1


In [48]:

# Retención por cohortes
# ======================

query_cohort_retention_final = '''
WITH creacion_cohortes_mensuales AS (    
    SELECT
    u.id_usuario,
    CAST(u.fecha_registro AS DATE) AS fecha_registro_date,
    DATE_TRUNC('month', CAST(u.fecha_registro AS DATE)) AS cohortes_mensuales,
    CAST(ua.fecha_actividad AS DATE) AS fecha_actividad_date,
    ua.dias_despues_registro,
    ua.activo
    FROM users u
    LEFT JOIN user_activity ua ON u.id_usuario = ua.id_usuario
    WHERE ua.activo = 1
    ),


retencion_semanal AS (
    SELECT
        cohortes_mensuales,
        COUNT(DISTINCT id_usuario) AS usuarios_iniciales,
        COUNT(DISTINCT CASE WHEN dias_despues_registro >= 1 AND dias_despues_registro <= 7 THEN id_usuario END) AS retenido_w1,
        COUNT(DISTINCT CASE WHEN dias_despues_registro >= 8 AND dias_despues_registro <= 14 THEN id_usuario END) AS retenido_w2,
        COUNT(DISTINCT CASE WHEN dias_despues_registro >= 15 AND dias_despues_registro <= 21 THEN id_usuario END) AS retenido_w3
        FROM creacion_cohortes_mensuales
        GROUP BY cohortes_mensuales
)

SELECT 
    TO_CHAR(cohortes_mensuales, 'YYYY-MM') AS cohorte,
    usuarios_iniciales,
    retenido_w1,
    ROUND(retenido_w1::numeric/NULLIF(usuarios_iniciales, 0)*100, 2) AS semana_1_pct,
    retenido_w2,
    ROUND(retenido_w2::numeric/NULLIF(usuarios_iniciales, 0)*100, 2) AS semana_2_pct,
    retenido_w3,
    ROUND(retenido_w3::numeric/NULLIF(usuarios_iniciales, 0)*100, 2) AS semana_3_pct
FROM retencion_semanal;


'''

# Ejecutar la consulta
cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)
cohorte_final


,cohorte,usuarios_iniciales,retenido_w1,semana_1_pct,retenido_w2,semana_2_pct,retenido_w3,semana_3_pct
0,2025-01,1381,697,50.47,668,48.37,656,47.50
1,2025-02,1255,611,48.69,609,48.53,635,50.60
2,2025-03,1428,677,47.41,705,49.37,690,48.32
3,2025-04,1394,680,48.78,697,50.00,663,47.56
4,2025-05,1446,695,48.06,676,46.75,706,48.82


<div class="alert alert-block alert-danger"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Esta sección necesita corrección antes de poder considerarse validada. La idea general de cohortes está bien orientada porque partes de fecha_registro para crear cohortes_mensuales, pero hay dos problemas técnicos importantes en la lógica de retención.

Primero, en la CTE inicial filtras con WHERE ua.activo = 1. Eso hace que usuarios_iniciales se calcule solo sobre usuarios que tuvieron actividad activa, no sobre todos los usuarios registrados en la cohorte. En retención, el denominador debe representar la base inicial de usuarios de cada cohorte; si excluimos a los inactivos desde el inicio, la retención queda inflada.

Segundo, estás calculando retenido_w1, retenido_w2 y retenido_w3 con rangos exactos de días (1 a 7, 8 a 14, 15 a 21). La rúbrica espera una lógica semanal acumulada, usando una semana derivada de dias_despues_registro / 7 y condiciones tipo semana mayor o igual a 1, 2 y 3. Si usas semanas exactas, estás midiendo actividad puntual en cada rango, no permanencia o regreso acumulado.

Para corregirlo, revisa la base de usuarios iniciales desde users, calcula la semana a partir de dias_despues_registro y luego construye la retención acumulada sobre esa base. Este ajuste es clave porque la retención es una métrica sensible: un denominador incorrecto puede llevar al negocio a pensar que los usuarios regresan más de lo que realmente ocurre.

</div>

---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado** 
4. **Interpretar el resultado**  

---
Hipótesis estadística
   - **H₀ (Hipótesis nula):** La proporción de conversión del grupo control es igual a la proporción de conversión del grupo tratamiento.
   - **H₁ (Hipótesis alternativa):** La proporción de conversión del grupo control es diferente a la proporción de conversión del grupo tratamiento.
   
**Test estadístico:** Prueba Z para comparación de dos proporciones independientes  
**Nivel de significancia alpha:** alpha = 0.05

In [49]:
experimentos = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv')

In [50]:
experimentos_clean = experimentos.copy()

In [51]:
experimentos_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_usuario       10000 non-null  object 
 1   variante         10000 non-null  object 
 2   convirtio        10000 non-null  int64  
 3   dispositivo      10000 non-null  object 
 4   pais             10000 non-null  object 
 5   duracion_sesion  10000 non-null  float64
 6   timestamp        10000 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 547.0+ KB


In [52]:
(experimentos_clean.groupby('variante')['convirtio'].mean())*100

variante
control        15.689829
tratamiento    16.285998
Name: convirtio, dtype: float64

In [53]:
conversiones = experimentos_clean.groupby('variante')['convirtio'].sum()
conversiones

variante
control        779
tratamiento    820
Name: convirtio, dtype: int64

In [54]:
total_usuarios = experimentos_clean.groupby('variante')['convirtio'].count()
total_usuarios

variante
control        4965
tratamiento    5035
Name: convirtio, dtype: int64

In [55]:
conversiones_grupo = [conversiones['control'], conversiones['tratamiento']]
conversiones_grupo

[779, 820]

In [56]:
observaciones = [total_usuarios['control'], total_usuarios['tratamiento']]
observaciones

[4965, 5035]

In [57]:
z_stat, p_value = proportions_ztest(conversiones_grupo, observaciones)

print(f'Estadistico ztest: {z_stat}')
print(f'Valor p: {p_value}')

Estadistico ztest: -0.8132782986429474
Valor p: 0.41605851639119995


In [58]:
alpha = 0.05

if p_value < alpha:
    print('Rechazamos la hipótesis nula: hay evidencia de una diferencia.')
else:
    print('No rechazamos la hipótesis nula: no hay evidencia suficiente de una diferencia.')

No rechazamos la hipótesis nula: no hay evidencia suficiente de una diferencia.


El grupo tratamiento presentó una tasa de conversión de 16.29%, ligeramente superior a la del grupo control (15.69%). Para determinar si esta diferencia es estadísticamente significativa, se aplicó una prueba Z para comparación de dos proporciones independientes con un nivel de significancia de 0.05. El resultado arrojó un estadístico Z de -0.8133 y un valor p de 0.4161. Debido a que el valor p es mayor que α = 0.05, no se rechaza la hipótesis nula. Por lo tanto, no existe evidencia estadística suficiente para afirmar que la modificación de la interfaz del checkout haya generado un impacto significativo en la tasa de conversión.

No se recomienda implementar el cambio únicamente por su efecto en la conversión, ya que la diferencia observada podría deberse al azar. Se sugiere continuar evaluando otras métricas o realizar nuevas pruebas para confirmar el impacto del cambio.

<div class="alert alert-block alert-success"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Muy buen trabajo en el test estadístico. Planteaste hipótesis nula y alternativa alineadas con la tasa de conversión, preparaste correctamente las conversiones y el total de usuarios por variante, aplicaste proportions_ztest y comparaste el p-valor contra alpha = 0.05.

La interpretación también es adecuada: como el p-valor es mayor a 0.05, no se rechaza la hipótesis nula y no hay evidencia estadística suficiente para afirmar que la nueva interfaz haya generado una diferencia significativa en conversión.

Me parece especialmente positivo que no hayas concluido que las tasas son “iguales” de forma absoluta, sino que usaste un lenguaje estadístico prudente. Esa precisión es importante en un contexto profesional, porque evita convertir ausencia de evidencia en una afirmación demasiado fuerte.

</div>

---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: Overview Ejecutivo
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

https://drive.google.com/drive/folders/1VzycaXodvapSALaRh65rihNzSUIuULyg?usp=sharing

<div class="alert alert-block alert-success"> <b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Muy buen trabajo con el dashboard, Rodrigo. A partir de las capturas se observa una estructura clara y bastante completa para una primera entrega: una página ejecutiva con KPIs principales, filtros por país y categoría, evolución mensual de ingresos y utilidad, acumulado del periodo, análisis por categoría y top de productos. Además, incluiste una segunda página de detalle o exploración con filtros, tarjeta de producto seleccionado, cantidad vendida y tabla de órdenes.

Esto cumple muy bien la idea de storytelling de negocio: primero muestras una visión general para entender el desempeño y luego permites profundizar en productos y pedidos. No se espera que el dashboard sea extremadamente sofisticado, pero el tuyo sí logra comunicar los resultados de manera clara y ordenada.

Como recomendación no bloqueante, revisa que la métrica llamada “Utilidad Total” tenga la misma definición que uses en el notebook. Si corresponde a utilidad bruta, el nombre debería reflejarlo; si corresponde a profit total según la rúbrica, entonces debe descontar marketing. Ese ajuste de consistencia hará que el dashboard sea más confiable para una audiencia ejecutiva.

</div>